Project.toml

```diff
@@ -0,0 +1,16 @@
+name = "CLFFT"
+uuid = "fe1fc2a7-b730-4eae-aba2-587af0173dc9"
+authors = []
+version = "0.1.0"
+
+[deps]
+BinDeps = "9e28174c-4ba2-5203-b857-d8d62c4213ee"
+BinaryProvider = "b99e7846-7c00-51b0-8f62-c81ae34c0232"
+Compat = "34da2185-b29b-5c13-b0c7-acf172513d20"
+FFTW = "7a1cc6ca-52ef-59f5-83cd-3a7055c09341"
+InfoZIP = "f4508453-b816-52ab-a864-26fc7f6211fc"
```

Comment:
This InfoZIP dependency here is strange. Are you sure that BinaryProvider.jl doesn't give you the choice of creating a Project.toml just for the build folder?

Reply:
@giordano appreciate if you can give some help here. We are trying to revive CLFFT.jl for Julia v1.6 and beyond and one thing that we are trying to do is use BinaryProvider.jl to download and install the binaries. Is this the correct way of doing it? What about these zip dep just to unpack the binary for Windows and save it somewhere?

Reply:
> Is this the correct way of doing it?

BinaryProvider has been unmaintained for the last 3-4 years, so I'd say definitely no.  Starting with it nowadays would be quite weird, and most importantly it won't work on new systems invented way after the BinaryProvider was still a thing, like Apple Silicon.

Reply:
> Oh I thought BinaryProvider.jl was the latest method to get binaries into place.

It was 4 years ago maybe.  Not after that.

> What is the new recommended method?

https://binarybuilder.org/

Reply:
We misunderstood the message then. We thought that BinaryBuilder.jl was used to build binaries to be used by BinaryProvider.jl in packages. The naming of the packages confused us basically.

We will try to run the BinaryBuilder.jl workflow (which is awesome BTW) to perform the cross-compilation.

Reply:
> We misunderstood the message then. We thought that BinaryBuilder.jl was used to build binaries to be used by BinaryProvider.jl in packages. The naming of the packages confused us basically.

No.  BinaryBuilder.jl is still used to build the binaries, but for the last 3 years it has been producing packages called "[JLLs](https://docs.binarybuilder.org/stable/jll/)" which replace entirely `BinaryProvider.jl`, which should ***not*** be used anywhere nowadays: https://julialang.org/blog/2019/11/artifacts/

Reply:
@giordano , I've successfully built OpenCL from their source code as follows:

```bash
git clone https://github.com/KhronosGroup/OpenCL-SDK
git submodule init
git submodule update
cmake .
make -j${proc}
make install
```

Now I am trying reproduce these steps with the BinaryBuilder.jl wizard. It asks if I need binary dependencies for the build, and I say yes, I need X11_jll. After selecting the dependency by typing "X11", the CMake system can't find it still:

![image](https://user-images.githubusercontent.com/3345261/192098803-18446a09-1ddc-4c8b-b421-1ad4853bcfd1.png)

Where should I set these PATH and LIB variables mentioned in the error message so that they are visible to the BinaryBuilder.jl build environment? Can I edit the CMakeLists.txt directly and BinaryBuilder.jl will record editions? Any help is appreciated.

My plan is  to provide a OpenCL_jll and then later provide a CLFFT_jll. Both of them have been built successfully on my local machine outside the BinaryBuilder.jl environment.

Reply:
1. do ***not*** use `X11_jll`, that's ***old*** and no one should use it
2. I presume (just guessing) you want instead `Xorg_libX11_jll`?
3. header files are in a separate package (that's how Xorg works, not our choice) called `Xorg_xorgproto_jll`

Reply:
Awesome! I searched for X11 in JuliaBinaryWrappers and didnt find the Xorg alternatives. I just need the lib (and possibly headers). Trying it now :)

Reply:
Do we have wrappers for the libraries and headers of OpenGL? It seems like the OpenCL build depends on it as well.

Reply:
![image](https://user-images.githubusercontent.com/3345261/192099560-323f9cec-9dd5-4851-bf52-b4af16b5c1d2.png)

Reply:
https://github.com/JuliaBinaryWrappers/Libglvnd_jll.jl#products?

Reply:
The X11 step passed successfully. Now I wonder what is necessary to make the OpenGL visible. Apparently every system comes with it pre-installed: https://www.khronos.org/opengl/wiki/Getting_Started#Downloading_OpenGL

Reply:
Found it! GLFW_jll :)

Reply:
@giordano I am almost there I think. Appreciate if you can help with this final CMake setup, which only occurs inside the BinaryBuilder.jl environment.

```julia
Would you like to resume the previous incomplete wizard run?
   Resume previous run
 > Start from scratch

    o      `.
   o*o      \'-_               00000000: 01111111  .
     \\      \;"".     ,;.--*  00000001: 01000101  E
      \\     ,\''--.--'/       00000002: 01001100  L
      :=\--<' `""  _   |       00000003: 01000110  F
      ||\\     `" / ''--       00000004: 00000010  .
      `/_\\,-|    |            00000005: 00000001  .
          \\/     L
           \\ ,'   \
         _/ L'   `  \          Join us in the #binarybuilder channel on the
        /  /    /   /          community slack: https://julialang.org/slack/
       /  /    |    \
      "_''--_-''---__=;        https://github.com/JuliaPackaging/BinaryBuilder.jl

Welcome to the BinaryBuilder wizard.  We'll get you set up in no time.

                        # Step 1: Select your platforms

Make a platform selection
 > All Supported Platforms
   Select by Operating System
   Fully Custom Platform Choice

Apple restricts distribution and usage of the macOS SDK, a necessary
component to build software for macOS targets.  Please read the Apple
and Xcode SDK agreement for more information on the restrictions and
legal terms you agree to when using the SDK to build software for Apple
operating systems: https://images.apple.com/legal/sla/docs/xcode.pdf.
Would you like to download and use the macOS SDK? [y/N]: n

Removing MacOS from the list of platforms...

                        # Step 2a: Obtain the source code

Please enter a URL (git repository or compressed archive) containing the source code to build or `N` to stop:
> https://github.com/KhronosGroup/OpenCL-SDK

The entered URL has been canonicalized to
https://github.com/KhronosGroup/OpenCL-SDK.git

┌ Info: Using cached git repository
│   url = "https://github.com/KhronosGroup/OpenCL-SDK.git"
└   repo_path = "/home/juliohm/.julia/packages/BinaryBuilderBase/h1spZ/deps/downloads/clones/OpenCL-SDK.git-1b21cf835d6531845310930c8dbc33ac8f59f3e1ee66fb5a0279302171687691"
You have selected a git repository. Please enter a branch, commit or tag to use.
Please note that for reproducibility, the exact commit will be recorded, 
so updates to the remote resource will not be used automatically; 
you will have to manually update the recorded commit.
> v2022.09.23

Recorded as 306a1837f28881730dd3f495a6a438942d84c2b4

Would you like to download additional sources?  [y/N]: n

                        # Step 2b: Obtain binary dependencies (if any)

Do you require any (binary) dependencies?  [y/N]: y
Enter JLL package name:
> Xorg_libX11

    Updating registry at `~/.julia/registries/General.toml`
Would you like to provide additional dependencies?  [y/N]: y
Enter JLL package name:
> Xorg_xorgproto

Would you like to provide additional dependencies?  [y/N]: y
Enter JLL package name:
> GLFW

Would you like to provide additional dependencies?  [y/N]: y
Enter JLL package name:
> GLU

Would you like to provide additional dependencies?  [y/N]: n

[ Info: Updating bare Yggdrasil clone in deps/Yggdrasil...
Enter a name for this project.  This will be used for filenames:
> OpenCL

Enter a version number for this project:
> v2022.09.23

Do you want to customize the set of compilers? [y/N]: n
                        # Step 3: Build for Linux x86_64 {libc=glibc}

You will now be dropped into the cross-compilation environment.
Please compile the package. Your initial compilation target is x86_64-linux-gnu
The $prefix environment variable contains the target directory.
Once you are done, exit by typing `exit` or `^D`

    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
   Installed GLU_jll ─ v9.0.1+0
  Downloaded artifact: GLU
    Updating `/tmp/jl_zKrK3i/XLdK2Wgp/x86_64-linux-gnu-libgfortran3-cxx03/.project/Project.toml`
  [0656b61e] + GLFW_jll v3.3.8+0
  [bd17208b] + GLU_jll v9.0.1+0
  [4f6342f7] + Xorg_libX11_jll v1.6.9+4
  [c4d99508] + Xorg_xorgproto_jll v2019.2.0+2
    Updating `/tmp/jl_zKrK3i/XLdK2Wgp/x86_64-linux-gnu-libgfortran3-cxx03/.project/Manifest.toml`
  [34da2185] + Compat v2.2.1
  [692b3bcd] + JLLWrappers v1.4.1
  [21216c6a] + Preferences v1.3.0
  [0656b61e] + GLFW_jll v3.3.8+0
  [bd17208b] + GLU_jll v9.0.1+0
  [d4300ac3] + Libgcrypt_jll v1.8.7+0
  [7e76a0d4] + Libglvnd_jll v1.3.0+3
  [7add5ba3] + Libgpg_error_jll v1.42.0+0
  [94ce4f54] + Libiconv_jll v1.16.1+1
  [02c8fc9c] + XML2_jll v2.9.14+0
  [aed1982a] + XSLT_jll v1.1.34+0
  [4f6342f7] + Xorg_libX11_jll v1.6.9+4
  [0c0b7dd1] + Xorg_libXau_jll v1.0.9+4
  [935fb764] + Xorg_libXcursor_jll v1.2.0+4
  [a3789734] + Xorg_libXdmcp_jll v1.1.3+4
  [1082639a] + Xorg_libXext_jll v1.3.4+4
  [d091e8ba] + Xorg_libXfixes_jll v5.0.3+4
  [a51aa0fd] + Xorg_libXi_jll v1.7.10+4
  [d1454406] + Xorg_libXinerama_jll v1.1.4+4
  [ec84b674] + Xorg_libXrandr_jll v1.5.2+4
  [ea2f1a96] + Xorg_libXrender_jll v0.9.10+4
  [14d82f49] + Xorg_libpthread_stubs_jll v0.1.0+3
  [c7cfdc94] + Xorg_libxcb_jll v1.13.0+3
  [c4d99508] + Xorg_xorgproto_jll v2019.2.0+2
  [c5fb5394] + Xorg_xtrans_jll v1.4.0+3
  [0dad84c5] + ArgTools v1.1.1
  [56f22d72] + Artifacts v1.3.0
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8bb1440f] + DelimitedFiles
  [8ba89e20] + Distributed
  [f43a241f] + Downloads v1.6.0
  [7b1f6079] + FileWatching
  [b77e0a4c] + InteractiveUtils
  [b27032c2] + LibCURL v0.6.3
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [ca575930] + NetworkOptions v1.2.0
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA v0.5.7
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [fa267f1f] + TOML v1.0.3
  [a4e569a6] + Tar v1.9.2
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [deac9b47] + LibCURL_jll v7.84.0+1
  [29816b5a] + LibSSH2_jll v1.10.2+2
  [c8ffd9c3] + MbedTLS_jll v2.28.0+2
  [14a3606d] + MozillaCACerts_jll v2022.3.29+0
  [83775a58] + Zlib_jll v1.2.12+4
  [8e850b90] + libblastrampoline_jll v5.1.1+1
  [8e850ede] + nghttp2_jll v1.48.0+1
  [3f19e933] + p7zip_jll v17.4.0+1
You have the following contents in your working directory:
  - OpenCL-SDK
Hints:
  - OpenCL-SDK/CMakeLists.txt

    This file is likely input to CMake. The recommended options for CMake are

      cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DCMAKE_BUILD_TYPE=Release

    followed by `make` and `make install`. Since the prefix environment
    variable is set already, this will automatically perform the installation
    into the correct directory.


sandbox:${WORKSPACE}/srcdir # cd OpenCL-SDK
sandbox:${WORKSPACE}/srcdir/OpenCL-SDK # git submodule init
Submodule 'headers-cpp' (https://github.com/KhronosGroup/OpenCL-CLHPP.git) registered for path 'external/OpenCL-CLHPP'
Submodule 'headers' (https://github.com/KhronosGroup/OpenCL-Headers.git) registered for path 'external/OpenCL-Headers'
Submodule 'loader' (https://github.com/KhronosGroup/OpenCL-ICD-Loader.git) registered for path 'external/OpenCL-ICD-Loader'
sandbox:${WORKSPACE}/srcdir/OpenCL-SDK # git submodule update
Cloning into '/workspace/srcdir/OpenCL-SDK/external/OpenCL-CLHPP'...
Cloning into '/workspace/srcdir/OpenCL-SDK/external/OpenCL-Headers'...
Cloning into '/workspace/srcdir/OpenCL-SDK/external/OpenCL-ICD-Loader'...
Submodule path 'external/OpenCL-CLHPP': checked out '4a1157466afe72a87e8abc59537ef577534ccadf'
Submodule path 'external/OpenCL-Headers': checked out '5e2f8c66476bcf60eb8c2921807e34efb30294ee'
Submodule path 'external/OpenCL-ICD-Loader': checked out '4aa683775c4ee55594196c79def7a75103d75e4c'
sandbox:${WORKSPACE}/srcdir/OpenCL-SDK # cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DCMAKE_BUILD_TYPE=Release
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


Re-run cmake no build system arguments
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/bin/x86_64-linux-gnu-libgfortran3-cxx03/x86_64-linux-gnu-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/bin/x86_64-linux-gnu-libgfortran3-cxx03/x86_64-linux-gnu-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Looking for secure_getenv
-- Looking for secure_getenv - not found
-- Looking for __secure_getenv
-- Looking for __secure_getenv - found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- cargs (https://github.com/likle/cargs) not found. To self-host, set cargs_INCLUDE_PATH and cargs_LIBRARY to point to the headers and library respectively adding '-D cargs_INCLUDE_PATH=/path/to/cargs/include/dir -D cargs_LIBRARY/path/to/cargs/libcargs' to the cmake command. (missing: cargs_INCLUDE_PATH cargs_LIBRARY) 
-- Fetching cargs.
-- Adding cargs subproject: /workspace/srcdir/OpenCL-SDK/_deps/cargs-external-src
-- TCLAP (http://tclap.sourceforge.net/) not found. To self-host, set TCLAP_INCLUDE_PATH to point to the headers adding '-DTCLAP_INCLUDE_PATH=/path/to/tclap' to the cmake command. (missing: TCLAP_INCLUDE_PATH) 
-- Fetching TCLAP.
-- Found TCLAP: /workspace/srcdir/OpenCL-SDK/_deps/tclap-external-src/include  
-- Stb (https://github.com/nothings/stb) not found. To self-host, set Stb_INCLUDE_PATH to point to the headers adding '-D Stb_INCLUDE_PATH=/path/to/stb' to the cmake command. (missing: Stb_INCLUDE_PATH) 
-- Fetching Stb.
-- Found Stb: /workspace/srcdir/OpenCL-SDK/_deps/stb-external-src  
-- Found X11: /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/include   
-- Looking for XOpenDisplay in /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libX11.so;/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libXext.so
-- Looking for XOpenDisplay in /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libX11.so;/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Could NOT find glm (missing: glm_DIR)
-- Fetching glm.
-- Adding glm subproject: /workspace/srcdir/OpenCL-SDK/_deps/glm-external-src
CMake Warning (dev) at /usr/share/cmake/Modules/FindOpenGL.cmake:315 (message):
  Policy CMP0072 is not set: FindOpenGL prefers GLVND by default when
  available.  Run "cmake --help-policy CMP0072" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  FindOpenGL found both a legacy GL library:

    OPENGL_gl_LIBRARY: /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libGL.so

  and GLVND libraries for OpenGL and GLX:

    OPENGL_opengl_LIBRARY: /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libOpenGL.so
    OPENGL_glx_LIBRARY: /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libGLX.so

  OpenGL_GL_PREFERENCE has not been set to "GLVND" or "LEGACY", so for
  compatibility with CMake 3.10 and below the legacy GL library will be used.
Call Stack (most recent call first):
  cmake/Dependencies/OpenGL/OpenGL.cmake:1 (find_package)
  cmake/Dependencies.cmake:10 (include)
  CMakeLists.txt:49 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found OpenGL: /opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/local/lib/libOpenGL.so   
-- Could NOT find GLEW (missing: GLEW_INCLUDE_DIRS GLEW_LIBRARIES) 
-- Fetching GLEW.
-- Adding GLEW subproject: /workspace/srcdir/OpenCL-SDK/_deps/glew-external-src
CMake Warning (dev) at _deps/glew-external-src/CMakeLists.txt:2 (project):
  Policy CMP0048 is not set: project() command manages VERSION variables.
  Run "cmake --help-policy CMP0048" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The following variable(s) would be set to empty:

    PROJECT_VERSION
    PROJECT_VERSION_MAJOR
    PROJECT_VERSION_MINOR
    PROJECT_VERSION_PATCH
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find Freetype (missing: FREETYPE_LIBRARY FREETYPE_INCLUDE_DIRS) 
-- Fetching FreeType.
-- Adding FreeType subproject: /workspace/srcdir/OpenCL-SDK/_deps/freetype-external-src
CMake Warning (dev) at _deps/freetype-external-src/CMakeLists.txt:162 (project):
  Policy CMP0048 is not set: project() command manages VERSION variables.
  Run "cmake --help-policy CMP0048" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.

  The following variable(s) would be set to empty:

    PROJECT_VERSION
    PROJECT_VERSION_MAJOR
    PROJECT_VERSION_MINOR
    PROJECT_VERSION_PATCH
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at _deps/freetype-external-src/CMakeLists.txt:222 (message):
  In-source builds are not permitted! Make a separate folder for building,
  e.g.,

    cmake -E make_directory build
    cmake -E chdir build cmake ..

  Before that, remove the files created by this failed run with

    cmake -E remove CMakeCache.txt
    cmake -E remove_directory CMakeFiles


-- Configuring incomplete, errors occurred!
See also "/workspace/srcdir/OpenCL-SDK/CMakeFiles/CMakeOutput.log".
See also "/workspace/srcdir/OpenCL-SDK/CMakeFiles/CMakeError.log".
```

Reply:
Did you read the error message
```
CMake Error at _deps/freetype-external-src/CMakeLists.txt:222 (message):
  In-source builds are not permitted! Make a separate folder for building,
```
?

Before running CMake create a build directory with
```
mkdir build && cd build
```
and then run CMake with
```
cmake -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} -DCMAKE_BUILD_TYPE=Release ..
```

Reply:
@giordano the GLFW and GLU binary deps worked, but the error seems to be related to a CMake setup now on one of the submodules. I wonder where I should edit the CMake configs suggested in the error message.

Reply:
But that error message was triggered by a CMake call to a submodule of the main CMake build. I will try to create this global build directory and see if it works.

Reply:
Awesome. One more step solved. Now it is asking for the UDev library. Do we have a _jll for it already?

```julia
CMake Error at build/_deps/sfml-external-src/cmake/Modules/FindUDev.cmake:49 (MESSAGE):
  Could not find UDev library
Call Stack (most recent call first):
  build/_deps/sfml-external-src/cmake/Macros.cmake:279 (find_package)
  build/_deps/sfml-external-src/src/SFML/Window/CMakeLists.txt:276 (sfml_find_package)
```

Reply:
Probably not

Reply:
Ok, I will try to work on the udev_jll and then come back here.